src/mode.c

```diff
@@ -0,0 +1,106 @@
+#include <stdint.h>
+#include <stdbool.h>
+
+#include "rc.h"
+#include "param.h"
+#include "mavlink_util.h"
+
+#include <breezystm32/breezystm32.h>
+
+#include "mode.h"
+
+
+armed_state_t _armed_state;
+
+void init_mode(void)
+{
+  _armed_state = DISARMED;
+}
+
+void arm(void)
+{
+  _armed_state = ARMED;
+  LED1_ON;
+}
+
+void disarm(void)
+{
+  _armed_state = DISARMED;
+  LED1_OFF;
+}
+
+bool check_mode(uint32_t now)
+{
+  static uint32_t prev_time = 0;
+  static uint32_t time_sticks_have_been_in_arming_position = 0;
+
+  // see it has been at least 20 ms
+  uint32_t dt = now-prev_time;
+  if (dt < 20000)
+  {
+    return false;
+  }
+
+  // if it has, then do stuff
+  prev_time = now;
+
+  // check for arming switch
+  if (_params.values[PARAM_ARM_STICKS])
+  {
+    if (_armed_state == DISARMED)
+    {
+      // if left stick is down and to the left
+      if (pwmRead(_params.values[PARAM_RC_F_CHANNEL]) < _params.values[PARAM_RC_F_BOTTOM] +
+          _params.values[PARAM_ARM_THRESHOLD]
+          && pwmRead(_params.values[PARAM_RC_Z_CHANNEL]) > (_params.values[PARAM_RC_Z_CENTER]+_params.values[PARAM_RC_Z_RANGE]/2)
+          - _params.values[PARAM_ARM_THRESHOLD])
+      {
+        mavlink_send_named_value_int("check", 1);
+        time_sticks_have_been_in_arming_position += dt;
+      }
+      else
+      {
+        mavlink_send_named_value_int("check", 2);
+        time_sticks_have_been_in_arming_position = 0;
+      }
+      if (time_sticks_have_been_in_arming_position > 500000)
+      {
+        mavlink_send_named_value_int("check", 3);
+        arm();
+        time_sticks_have_been_in_arming_position = 0;
+      }
+    }
+    else // _armed_state is ARMED
+    {
+      // if left stick is down and to the right
+      if (pwmRead(_params.values[PARAM_RC_F_CHANNEL]) < _params.values[PARAM_RC_F_BOTTOM] +
+          _params.values[PARAM_ARM_THRESHOLD]
+          && pwmRead(_params.values[PARAM_RC_Z_CHANNEL]) < (_params.values[PARAM_RC_Z_CENTER]-_params.values[PARAM_RC_Z_RANGE]/2)
```

Comment:
Also, on the fixed_wing_test branch (where most of my work has been focused the last two days) I added the ability to switch the direction on the servo output.  Does it make sense to add the ability to switch input directions, or does that just make it confusing?